cider-dap/src/main.rs

```diff
@@ -0,0 +1,139 @@
+    #[argh(option, short = 'p', long = "port", default = "8080")]
+    /// port for the TCP server
+    port: u16,
+}
+fn read_path(path: &str) -> Result<PathBuf, String> {
+    Ok(PathBuf::from(path))
+}
+
+fn main() -> Result<(), MyAdapterError> {
+    let opts: Opts = argh::from_env();
+
+    if opts.is_multi_session {
+        eprintln!("running multi-session");
+        let listener = TcpListener::bind(("127.0.0.1", opts.port))?;
+        eprintln!("bound on port: {} ", opts.port);
+        let (stream, addr) = listener.accept()?;
+        println!("Accepted client on: {}", addr);
+        let read_stream = BufReader::new(stream.try_clone()?);
+        let write_stream = BufWriter::new(stream);
+        let mut server = Server::new(read_stream, write_stream);
+        multi_session_init(&mut server)?;
+    } else {
+        let path = opts.file.ok_or(MyAdapterError::MissingFile)?;
+        let file = File::open(path)?;
+        let adapter = MyAdapter::new(file);
+        eprintln!("running single-session");
+        let write = BufWriter::new(stdout());
+        let read = BufReader::new(stdin());
+        let server = Server::new(read, write);
+        // run_server(server, adapter)?;
+    }
+    eprintln!("exited run_Server");
+    Ok(())
+}
+
+fn multi_session_init<R, W>(server: &mut Server<R, W>) -> AdapterResult<()>
+where
+    R: Read,
+    W: Write,
+{
+    // handle the first request (Initialize)
+    let req = match server.poll_request()? {
+        Some(req) => req,
+        None => return Err(MyAdapterError::MissingCommandError),
+    };
+    match &req.command {
+        Command::Initialize(_) => {
+            let rsp =
+                req.success(ResponseBody::Initialize(types::Capabilities {
+                    ..Default::default()
+                }));
+            server.respond(rsp)?;
+            server.send_event(Event::Initialized)?;
+        }
+        _ => unreachable!(),
+    }
+
+    // handle the second request (Launch)
+    let req = match server.poll_request()? {
+        Some(req) => req,
+        None => return Err(MyAdapterError::MissingCommandError),
+    };
+
+    let program_path = if let Command::Launch(ref params) = &req.command {
+        if let Some(data) = &params.additional_data {
+            if let Some(program_path) = data.get("program") {
+                eprintln!("Program path: {}", program_path);
+                program_path
+                    .as_str()
+                    .ok_or(MyAdapterError::InvalidPathError)?
+            } else {
+                return Err(MyAdapterError::MissingFile);
+            }
+        } else {
+            return Err(MyAdapterError::MissingFile);
+        }
+    } else {
+        unreachable!();
+    };
+
+    // Open file using the extracted program path
+    let file = File::open(program_path)?;
+
+    // Construct the adapter
+    let adapter = MyAdapter::new(file);
+
+    // Now, run the server with the adapter
+    run_server(server, adapter)
```

Comment:
This is fine, though this nesting seems a bit unnecessary since you could just
return the adapter constructed in this function and run the server in the parent
context. Feels a bit off that the init then just runs the server.